### PR TITLE
Openlibrary integration

### PR DESCRIPTION
**Pull request by mnmallea**
## Summary
- Added `GET /book_information` endpoint to search books by ISBN.
- Added BookInformationController
- Created OpenlibraryService to fetch books in Openlibrary API

## Trello card
[Here](https://trello.com/c/ClTd8vzQ/35-integraci%C3%B3n-api-externa)

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -31,6 +31,7 @@ gem 'sidekiq', '~>5.2.7'
 gem 'validate_url', '~> 1.0.8', git: 'https://github.com/perfectline/validates_url.git',
                                 branch: 'master'
 gem 'wor-paginate'
+gem 'httparty'
```
Comment (lucasVoboril) [inline Gemfile]: Tag gem version
```suggestion
gem 'httparty', '~> 0.17.0'
```

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -14,6 +14,11 @@ def configure_permitted_parameters
     devise_parameter_sanitizer.permit(:sign_up, keys: %i[first_name last_name locale])
   end
 
+  def render_errors(message, status)
```
Comment (lucasVoboril) [inline app/controllers/application_controller.rb]: This change comes from the previous PR. You should change your base branch :wink:

File: app/controllers/book_information_controller.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+class BookInformationController < ApplicationController
```
Comment (lucasVoboril) [inline app/controllers/book_information_controller.rb]: I would not create this controller.
Usually, the controllers are associated with models, and we haven't got BookInformation model.

File: app/controllers/book_information_controller.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  def show
```
Comment (lucasVoboril) [inline app/controllers/book_information_controller.rb]: We can move this method to Books controller.
`Books#open_library_information`

File: app/controllers/book_information_controller.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+    render json: OpenlibraryService.new.search_by_isbn(isbn), status: :ok
+  rescue OpenlibraryService::BookNotFoundError => e
+    render_errors(e.message, :not_found)
+  end
+
+  private
+
+  def isbn
+    @isbn ||= params[:isbn]
```
Comment (lucasVoboril) [inline app/controllers/book_information_controller.rb]: :clap: Nice [memorize](https://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/)

File: app/controllers/book_information_controller.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  def valid_isbn?
+    isbn&.contains_only_digits?
```
Comment (lucasVoboril) [inline app/controllers/book_information_controller.rb]: :clap: nice [safe navegation](http://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/)

File: config/initializers/core_exts/string.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class String
+  def contains_only_digits?
```
Comment (lucasVoboril) [inline config/initializers/core_exts/string.rb]: I prefer to call it `is_number?` [Check it](https://stackoverflow.com/questions/5661466/test-if-string-is-a-number-in-ruby-on-rails)

File: app/controllers/book_information_controller.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+class BookInformationController < ApplicationController
+  def show
+    return render_errors({ isbn: 'ISBN should be numeric' }, :bad_request) unless valid_isbn?
```
Comment (lucasVoboril) [inline app/controllers/book_information_controller.rb]: I prefer to use full_messages in errors, and not hashes.
```suggestion
    return render_errors('ISBN should be numeric' , :bad_request) unless valid_isbn?
```

File: app/services/openlibrary_service.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
+class OpenlibraryService
+  include HTTParty
+  base_uri 'https://openlibrary.org/api'
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb]: You should use an env variable to this base_uri

File: app/services/openlibrary_service.rb [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+  default_params format: :json, jscmd: 'data'
+
+  class BookNotFoundError < StandardError
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb]: We save CustomErrors in another file.
```rb
# app/errors/custom_errors.rb
module CustomErrors
  class BookNotFoundError < StandardError; end
end
```

And this
```rb
# app/controllers/application_controller.rb
include CustomErrors
```

File: app/services/openlibrary_service.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  def search_by_isbn(isbn)
+    isbn_string = "ISBN:#{isbn}"
+    options = { query: { bibkeys: isbn_string } }
+    response = self.class.get('/books', options)
+
+    return parse_response_book(response[isbn_string], isbn) if response.success?
+
+    raise BookNotFoundError, response.response
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb]:
```suggestion
    raise CustomErrors::BookNotFoundError, response.response
```

File: spec/controllers/book_information_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+describe BookInformationController do
```
Comment (lucasVoboril) [inline spec/controllers/book_information_controller_spec.rb]: We are going to test Service in another PR, and we don't neer this Controller specs :wink:

File: config/routes.rb [hunk 13]
```diff
@@ -8,5 +8,7 @@
     resources :rents, only: %i[index create]
   end
   resources :books, only: %i[index show]
+  get '/book_information', to: 'book_information#show'
```
Comment (lucasVoboril) [inline config/routes.rb]: I prefer use `open_library_information` endpoint in books controller. You can check differences between resources, collection and member [in this link](https://carlossanchezperez.wordpress.com/2013/07/21/definir-rutas-en-una-aplicacion-de-rails-con-new-resources-restful-route-customizations/) 
```rb
resources :books, only: %i[index show] do
  collection do
    :open_library_information
  end
end
```

File: app/services/openlibrary_service.rb [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+  def parse_response_book(book, isbn)
+    {
+      isbn: isbn,
+      title: book['title'],
+      subtitle: book['subtitle'],
+      number_of_pages: book['number_of_pages'],
+      authors: book['authors'].map { |author| author['name'] }
+    }
+  rescue StandardError
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb]: Why you are doing this rescue?

**Review by lucasVoboril — CHANGES_REQUESTED** [review]

File: app/controllers/book_information_controller.rb (hunk 8 above)
Comment (lucasVoboril) [reply app/controllers/book_information_controller.rb]: Also, if you use hashes errors:
```suggestion
    return render_errors({ isbn: 'should be numeric' }, :bad_request) unless valid_isbn?
```

File: app/services/openlibrary_service.rb (hunk 14 above)
Comment (mnmallea) [reply app/services/openlibrary_service.rb]: If `book['authors']` is `nil` or not an array it throws NoMethodError

Comment (mnmallea) [reply app/services/openlibrary_service.rb]:
```ruby
def parse_response_book(book, isbn)
    raise CustomErrors::BookNotFoundError, 'Invalid response book' unless
      book&.include_all?('title', 'subtitle', 'number_of_pages', 'authors') &&
      book['authors'].is_a?(Array)

    {
      isbn: isbn,
      title: book['title'],
      subtitle: book['subtitle'],
      number_of_pages: book['number_of_pages'],
      authors: book['authors'].map { |author| author['name'] }
    }
  end
```

File: app/services/openlibrary_service.rb [hunk 15]
```diff
@@ -0,0 +1,34 @@
+require 'set'
+
+class OpenlibraryService
+  include HTTParty
+  base_uri ENV['OPEN_LIBRARY_URI']
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb]:
```suggestion
  base_uri Rails.application.secrets.open_library_uri
```

File: app/services/openlibrary_service.rb [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    return parse_response_book(response[isbn_string], isbn) if response.success?
+
+    raise CustomErrors::BookNotFoundError, response.response
+  end
+
+  private
+
+  def parse_response_book(book, isbn)
+    raise CustomErrors::BookNotFoundError, 'Invalid response book' unless
+      book&.include_all?('title', 'subtitle', 'number_of_pages', 'authors') &&
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb]: When evaluate conditions, pay attention to `logical weight`. 
Ruby is lazy, and in this condition `A && B` if (A is false), then B will never be evaluated.
`logical weight` says: if you have 2 conditions, use the most "descartable" option first.
In this case, it's more easy to ask `book['authors'].is_a?(Array)` than for another condition.

Comment (lucasVoboril) [inline app/services/openlibrary_service.rb]: `include_all?` Nice condition, but it's easier if you use: 
```rb
(%w[title subtitle number_of_pages authors] - book.keys).empty?
```

File: app/services/openlibrary_service.rb [hunk 17]
```diff
@@ -0,0 +1,34 @@
+require 'set'
+
+class OpenlibraryService
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb:1]: Functionally, this service it's ok. But `search_by_isbn(isbn)` and `parse_response_book(book, isbn)` are too long methods. 
We should delegate if to many methods [check long method refactoring](https://sourcemaking.com/refactoring/smells/long-method)

File: app/services/openlibrary_service.rb [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+  include HTTParty
+  base_uri ENV['OPEN_LIBRARY_URI']
+
+  default_params format: :json, jscmd: 'data'
+
+  def search_by_isbn(isbn)
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb:7]:
```rb
def search_by_isbn(isbn)
  @isbn = isbn
  parse_book(open_library_response)
end

def open_library_response
  response = self.class.get('/books', query: { bibkeys: "ISBN:#{@isbn}" })
  raise CustomErrors::BookNotFoundError, response.response unless response.success?

  response
end
```

File: app/services/openlibrary_service.rb [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+    response = self.class.get('/books', options)
+
+    return parse_response_book(response[isbn_string], isbn) if response.success?
+
+    raise CustomErrors::BookNotFoundError, response.response
+  end
+
+  private
+
+  def parse_response_book(book, isbn)
```
Comment (lucasVoboril) [inline app/services/openlibrary_service.rb]:
```rb
  def parse_book(response)
    @book = response["ISBN:#{@isbn}"]
    valid_book_attributes? ? parsed_book_data : raise_invalid_response_book
  end

  def valid_book_attributes?
    searched_words = %w[title subtitle number_of_pages authors]
    @book.present? && @book['authors'].is_a?(Array) && (searched_words - @book.keys).empty?
  end

  def parsed_book_data
    {
      isbn: @isbn,
      title: @book['title'],
      subtitle: @book['subtitle'],
      number_of_pages: @book['number_of_pages'],
      authors: @book['authors'].map { |author| author['name'] }
    }
  end

  def raise_invalid_response_book
    raise CustomErrors::BookNotFoundError, 'Invalid response book'
  end
```

**Review by lucasVoboril — CHANGES_REQUESTED** [review]

**Review by lucasVoboril — APPROVED** [review]

**Pull request merged**